tests/unit/io/test_connection.py

```diff
@@ -183,3 +184,23 @@ def raise_dbapi_error_on_first_call(sql, bindings):
 
         self.assertNotEquals(connection, lore.io.main._connection)
         self.assertEquals(result[0][0], 1)
+
+    def test_reconnect_and_retry_on_expired_connection(self):
+        original_execute = lore.io.main._connection.execute
+
+        def raise_snowflake_programming_error_on_first_call(sql, bindings):
+            lore.io.main._connection.execute = original_execute
+            e = lore.io.connection.SnowflakeProgrammingError('Authentication token has expired.  The user must authenticate again')
+            raise sqlalchemy.exc.DBAPIError('select 1', [], e, True)
+
+        exceptions = lore.env.STDOUT_EXCEPTIONS
+        lore.env.STDOUT_EXCEPTIONS = False
+        connection = lore.io.main._connection
+        lore.io.main._connection.execute = raise_snowflake_programming_error_on_first_call
+
+        result = lore.io.main.select(sql='select 1')
```

Comment:
@montanalow should i specify connection to snowflake (lore.io.analysis._connection) here ? My understanding is that it shouldn't matter because anyways we are just throwing some error in response to db query.